rpc/rpcserver/server.go

```diff
@@ -2314,3 +2328,125 @@ func (s *messageVerificationServer) VerifyMessage(ctx context.Context, req *pb.V
+
+	for i, txIn := range mtx.TxIn {
+		// The disassembled string will contain [error] inline
+		// if the script doesn't fully parse, so ignore the
+		// error here.
+		disbuf, _ := txscript.DisasmString(txIn.SignatureScript)
+
+		inputs[i] = &pb.DecodedTransaction_Input{
+			PreviousTransactionHash:  txIn.PreviousOutPoint.Hash[:],
+			PreviousTransactionIndex: txIn.PreviousOutPoint.Index,
+			Tree:               pb.DecodedTransaction_Input_TreeType(txIn.PreviousOutPoint.Tree),
+			Sequence:           txIn.Sequence,
+			AmountIn:           txIn.ValueIn,
+			BlockHeight:        txIn.BlockHeight,
+			BlockIndex:         txIn.BlockIndex,
+			SignatureScript:    txIn.SignatureScript,
+			SignatureScriptAsm: disbuf,
+		}
+	}
+
+	return inputs
+}
+
+func marshalDecodedTxOutputs(mtx *wire.MsgTx, chainParams *chaincfg.Params) []*pb.DecodedTransaction_Output {
+	outputs := make([]*pb.DecodedTransaction_Output, len(mtx.TxOut))
+	txType := stake.DetermineTxType(mtx)
+
+	for i, v := range mtx.TxOut {
+		// The disassembled string will contain [error] inline if the
+		// script doesn't fully parse, so ignore the error here.
+		disbuf, _ := txscript.DisasmString(v.PkScript)
+
+		// Attempt to extract addresses from the public key script.  In
+		// the case of stake submission transactions, the odd outputs
+		// contain a commitment address, so detect that case
+		// accordingly.
+		var addrs []dcrutil.Address
+		var encodedAddrs []string
+		var scriptClass txscript.ScriptClass
+		var reqSigs int
+		var commitAmt *dcrutil.Amount
+		if (txType == stake.TxTypeSStx) && (stake.IsStakeSubmissionTxOut(i)) {
+			scriptClass = txscript.StakeSubmissionTy
+			addr, err := stake.AddrFromSStxPkScrCommitment(v.PkScript,
+				chainParams)
+			if err != nil {
+				encodedAddrs = []string{fmt.Sprintf(
+					"[error] failed to decode ticket "+
+						"commitment addr output for tx hash "+
+						"%v, output idx %v", mtx.TxHash(), i)}
+			} else {
+				encodedAddrs = []string{addr.EncodeAddress()}
+			}
+			amt, err := stake.AmountFromSStxPkScrCommitment(v.PkScript)
+			if err != nil {
+				commitAmt = &amt
+			}
+		} else {
+			// Ignore the error here since an error means the script
+			// couldn't parse and there is no additional information
+			// about it anyways.
+			scriptClass, addrs, reqSigs, _ = txscript.ExtractPkScriptAddrs(
+				v.Version, v.PkScript, chainParams)
+			encodedAddrs := make([]string, len(addrs))
+			for j, addr := range addrs {
+				encodedAddrs[j] = addr.EncodeAddress()
+			}
+		}
+
+		outputs[i] = &pb.DecodedTransaction_Output{
+			Index:              uint32(i),
+			Value:              v.Value,
+			Version:            int32(v.Version),
+			Addresses:          encodedAddrs,
+			Script:             v.PkScript,
+			ScriptAsm:          disbuf,
+			ScriptClass:        pb.DecodedTransaction_Output_ScriptClass(scriptClass),
+			RequiredSignatures: int32(reqSigs),
+		}
+		if commitAmt != nil {
+			outputs[i].CommitmentAmount = int64(*commitAmt)
```

Comment:
Looking at the code that generates the commitment output ([here](https://github.com/decred/dcrd/blob/master/blockchain/chaingen/generator.go#L479)) the value of output is always 0, so maybe we could also collapse this field into the output value field.

Then a consumer of the api could use the script class (`stakeSubmissionTy`) to determine that the output value is actually a commitment instead of the value.

On the other hand, this breaks the assumption that the output's value field in the decoded structure is always equal to the value of the corresponding field in the serialized transaction, so I should maybe leave this alone.